zeq/publisher.cpp

```diff
@@ -93,21 +94,46 @@ class Publisher
         return true;
     }
 
-    const std::string& getAddress() const { return _address; }
+    std::string getAddress() const
+    {
+        return uri.getHost() + ":" + std::to_string( uint32_t( uri.getPort( )));
+    }
+
+    lunchbox::URI uri;
 
 private:
-    void _initService( std::string host, uint16_t port )
+    void _initService( const uint32_t announceMode )
     {
-        _initAddress( host, port );
-        if( lunchbox::Servus::isAvailable( ))
-            _service.announce( port, _address );
+        _initAddress();
+        if( !(announceMode&ANNOUNCE_ZEROCONF) )
+            return;
+
+        const bool required = (announceMode & ANNOUNCE_REQUIRED);
+        if( !lunchbox::Servus::isAvailable( ))
+        {
+            if( required )
+                LBTHROW( std::runtime_error(
+                             "No zeroconf implementation available" ));
+            return;
+        }
+
+        const lunchbox::Servus::Result& result = _service.announce(
+            uri.getPort(), getAddress( ));
+
+        if( required && !result )
+        {
+            LBTHROW( std::runtime_error( "Zeroconf announce failed: " +
+                                         result.getString( )));
+        }
     }
 
-    void _initAddress( std::string& host, uint16_t& port )
```

Comment:
The changes you've made in this file, and particularly this function, are going to be very conflicting with hernando:optional_lunchbox. I'd be easier if Publisher::getURI returned a std::string.